python/ray/data/_internal/execution/interfaces/op_runtime_metrics.py

```diff
@@ -91,17 +91,14 @@ class OpRuntimeMetrics:
     obj_store_mem_freed: int = field(
         default=0, metadata={"map_only": True, "export_metric": True}
     )
-    # Current memory size in the object store.
-    obj_store_mem_cur: int = field(
-        default=0, metadata={"map_only": True, "export_metric": True}
-    )
-    # Peak memory size in the object store.
-    obj_store_mem_peak: int = field(default=0, metadata={"map_only": True})
     # Spilled memory size in the object store.
     obj_store_mem_spilled: int = field(
         default=0, metadata={"map_only": True, "export_metric": True}
     )
 
+    obj_store_mem_internal_inqueue: int = 0
+    obj_store_mem_internal_outqueue: int = 0
```

Comment:
These metrics currently only work for map ops. I'm wondering if we can also extend them to also support other ops. So 1) we can report them to the dashboard for all ops; 2) we don't have 2 places reporting memory metrics (here and internal_inqueue_memory_usage).

The `on_input_received` function is already called on all ops. I think we'll also need to add other callback APIs like `on_input_freed`, `on_output_generated`, etc.

Reply:
Yeah, I thought about that too. 

`OpRuntimeMetrics` is tightly coupled with `MapOperator`, and this PR introduces a layer of indirection with `PhysicalOperator.*_memory_usage()` and `OpRuntimeMetrics.obj_store_mem_*`.

One concern I have is that if we rely on metrics directly, we need to ensure that they're updated correctly for all operators (for example, we need to make sure all operators call `on_output_generated`, or else `obj_store_mem_internal_outqueue` will be incorrect). But, I'm not sure how we can enforce that with the `PhysicalOperator` interface.

Reply:
Documenting tradeoffs from offline discussion:

**Disadvantages of adding methods to PhysicalOperator:**
- OpRuntimeMetrics is tightly coupled with MapOperator
- Layer of indirection for MapOperator
- Multiple abstractions for memory reporting

**Disadvantages of using OpRuntimeMetrics only:**

- Updating metrics is implicitly required (not part of interface)

Given these tradeoffs, I've opened a new PR with the second approach https://github.com/ray-project/ray/pull/42930/files